rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+    #[default]
+    Air = 0x00,
+    Water = 0x01,
+    Grass = 0x02,
+    Slate = 0x03,
+    Concrete = 0x04,
+    Brick = 0x05,
+    Sand = 0x06,
+    WoodPlanks = 0x07,
+    Rock = 0x08,
+    Glacier = 0x09,
+    Snow = 0x0A,
+    Sandstone = 0x0B,
+    Mud = 0x0C,
+    Basalt = 0x0D,
+    Ground = 0x0E,
+    CrackedLava = 0x0F,
+    Asphalt = 0x10,
+    Cobblestone = 0x11,
+    Ice = 0x12,
+    LeafyGrass = 0x13,
+    Salt = 0x14,
+    Limestone = 0x15,
+    Pavement = 0x16,
+}
+
+impl TryFrom<u8> for Material {
+    type Error = InvalidMaterialError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        use Material::*;
+
+        Ok(match value {
+            0x00 => Air,
+            0x01 => Water,
+            0x02 => Grass,
+            0x03 => Slate,
+            0x04 => Concrete,
+            0x05 => Brick,
+            0x06 => Sand,
+            0x07 => WoodPlanks,
+            0x08 => Rock,
+            0x09 => Glacier,
+            0x0A => Snow,
+            0x0B => Sandstone,
+            0x0C => Mud,
+            0x0D => Basalt,
+            0x0E => Ground,
+            0x0F => CrackedLava,
+            0x10 => Asphalt,
+            0x11 => Cobblestone,
+            0x12 => Ice,
+            0x13 => LeafyGrass,
+            0x14 => Salt,
+            0x15 => Limestone,
+            0x16 => Pavement,
+            _ => return Err(InvalidMaterialError(value)),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidMaterialError(u8);
+
+impl std::error::Error for InvalidMaterialError {}
+
+impl fmt::Display for InvalidMaterialError {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "Invalid binary material value {:x?}", self.0)
+    }
+}
+
+pub trait TerrainSerializer {
+    fn encode(&self) -> Vec<u8>;
+    //fn decode(&self) -> Self;
+}
+
+/// A container for a voxel of terrain, used in the `Chunk` object.
+#[derive(Debug, Default, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Voxel {
+    solid_occupancy: f32,
+    water_occupancy: f32,
+    material: Material,
+}
+
+impl Voxel {
+    /// Constructs a new `Voxel` with a material and occupancy percentage.
+    /// Equivalent to data writeable from Roblox's `Terrain:WriteVoxels`.
+    /// Occupancy values are between `0.0` and `1.0`, as a percentage of the voxel.
+    pub fn new(material: Material, solid_occupancy: f32) -> Self {
+        let mut voxel = Self {
+            material,
+            ..Default::default()
+        };
+        voxel.set_occupancy(solid_occupancy, 0.0);
+
+        voxel
+    }
+
+    /// Constructs a new `Voxel` with a material, solid occupancy, and water
+    /// occupancy percentage. Takes advantage of the recently-released
+    /// Shorelines feature. Equivalent to data writeable from Roblox's `Terrain:WriteVoxelChannels`.
+    /// Occupancy values are between `0.0` and `1.0`, as a percentage of the voxel.
```

Comment:
```suggestion
    /// Constructs a new `Voxel` with a material, solid occupancy, and water
    /// occupancy percentage.
    /// Equivalent to data writeable from Roblox's `Terrain:WriteVoxelChannels`.
    /// Occupancy values are between `0.0` and `1.0`, as a percentage of the voxel.
```
This comment is going to be here theoretically forever so we should avoid statements like 'recently-released'.